# بسم الله الرحمن الرحيم

In [1]:
# import pandas as pd 
# import numpy as np 
# from matplotlib import pyplot as plt 
# import seaborn as sns 
# plt.style.use('default')
# #-------------------------------------------------------------------------------------
# from sklearn.utils import resample 
# from sklearn import preprocessing
# from sklearn.linear_model import LogisticRegression
# from sklearn.model_selection import train_test_split
# from sklearn.ensemble import RandomForestClassifier
# from sklearn.tree import DecisionTreeClassifier
# from sklearn.svm import SVC
# from sklearn.pipeline import make_pipeline
# from sklearn.metrics import confusion_matrix
# from sklearn.model_selection import cross_validate

# from sklearn.naive_bayes import GaussianNB
# from sklearn.metrics import confusion_matrix
# from sklearn.metrics import classification_report
# from sklearn.metrics import precision_score
# from sklearn.metrics import recall_score
# from sklearn.pipeline import Pipeline 

# import tensorflow_addons as tfa
# import tensorflow as tf 
# from tensorflow import keras 
# from tensorflow.keras import layers 
# from focal_loss import SparseCategoricalFocalLoss

# #-------------------------------------------------------------------------------------

# from keras.callbacks import EarlyStopping
# from keras.callbacks import ModelCheckpoint
# #-------------------------------------------------------------------------------------
# from imblearn.over_sampling import SMOTE
# from utils import *
# %matplotlib inline

In [3]:
from utils import *
import pandas as pd 
from sklearn.model_selection import train_test_split
import tensorflow as tf 
import tensorflow.keras 
import pickle 
import seaborn as sns 
from matplotlib import pyplot as plt 


## Reading the Dataset, cleaning , preparing it for the training 

In [4]:
df = pd.read_csv("dataset/body_level_classification_train.csv")

In [5]:
df_original= cleaning_data(df).copy()

In [6]:
all_class_train,all_class_test=data_spliting(df_original,1/3)

In [7]:
X_train=all_class_train.drop(["Body_Level"],axis=1).to_numpy()
Y_train=all_class_train["Body_Level"].to_numpy()
X_test =all_class_test.drop(["Body_Level"],axis=1).to_numpy()
Y_test =all_class_test["Body_Level"].to_numpy()

In [8]:
X_train, scaler= data_scaling(X_train)
with open(f"scaler","wb") as fp : 
    pickle.dump(scaler,fp)

In [9]:
X_val, X_test, Y_val, Y_test = train_test_split(X_test
                                                     , Y_test, test_size=0.5, random_state=42)

In [10]:
X_val = scaler.transform(X_val)
X_test=scaler.transform(X_test)

In [11]:
model_one_layer  = keras.Sequential([keras.Input(shape=(23)),
                           layers.Dense(4,activation='relu'),
                           
                        layers.Dense(4,activation='softmax')])
model_one_layer.compile(loss=SparseCategoricalFocalLoss(from_logits=False,gamma=2),
          optimizer=keras.optimizers.Adam(learning_rate=0.01),
          metrics=["accuracy"])

2023-05-15 10:18:10.250828: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2023-05-15 10:18:10.250856: W tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)
2023-05-15 10:18:10.250882: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (brownboyComputer): /proc/driver/nvidia/version does not exist
2023-05-15 10:18:10.251243: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [12]:
# model=build_one_layer_model(2,0.01)
model=model_one_layer

In [13]:
history = model.fit(X_train,Y_train,validation_data=[X_val,Y_val] 
             ,batch_size=32, epochs=300 , verbose =1, 
                       callbacks=[CustomCallback("NNModels/One_layer_before_sampling/before_sampling.h5")])

Epoch 1/300
31/31 [==============================] - 1s 9ms/step - loss: 0.6570 - accuracy: 0.5122 - val_loss: 0.5363 - val_accuracy: 0.5592
Epoch 2/300
31/31 [==============================] - ETA: 0s - loss: 0.4447 - accuracy: 0.6207
Yes You are here 0.6206896305084229 0.6081632375717163
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
......dense_1
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
.........5
.........6
.........7
.........8
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5                                   2023-05-15 10:18:15        21392
config.json                                    2023-05-15 10:18:15      

File Name                                             Modified             Size
variables.h5                                   2023-05-15 10:18:17        21392
config.json                                    2023-05-15 10:18:17         2007
metadata.json                                  2023-05-15 10:18:17           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0529 - accuracy: 0.9523 - val_loss: 0.0502 - val_accuracy: 0.9551
Epoch 17/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0493 - accuracy: 0.9523 - val_loss: 0.0481 - val_accuracy: 0.9510
Epoch 18/300
29/31 [===========================>..] - ETA: 0s - loss: 0.0439 - accuracy: 0.9580
Yes You are here 0.9563894271850586 0.9591836929321289
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
......dense_1
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......me

metadata.json                                  2023-05-15 10:18:18           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0273 - accuracy: 0.9736 - val_loss: 0.0295 - val_accuracy: 0.9714
Epoch 27/300
 1/31 [..............................] - ETA: 0s - loss: 0.0237 - accuracy: 1.0000
Yes You are here 0.9776875972747803 0.9714285731315613
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
......dense_1
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
.........5
.........6
.........7
.........8
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5                                   2023-05-15 10:18:18        21392
con

metadata.json                                  2023-05-15 10:18:19           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0158 - accuracy: 0.9828 - val_loss: 0.0206 - val_accuracy: 0.9837
Epoch 39/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0171 - accuracy: 0.9858 - val_loss: 0.0187 - val_accuracy: 0.9796
Epoch 40/300
 1/31 [..............................] - ETA: 0s - loss: 0.0306 - accuracy: 0.9688
Yes You are here 0.9868153929710388 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
......dense_1
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
.........5
.........6
.........7
.........8
...vars
Keras model archive saving:
File Name            

31/31 [==============================] - 0s 4ms/step - loss: 0.0112 - accuracy: 0.9919 - val_loss: 0.0191 - val_accuracy: 0.9837
Epoch 50/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0104 - accuracy: 0.9919 - val_loss: 0.0214 - val_accuracy: 0.9837
Epoch 51/300
 1/31 [..............................] - ETA: 0s - loss: 0.0105 - accuracy: 1.0000
Yes You are here 0.9908722043037415 0.9877551198005676
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
......dense_1
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
.........5
.........6
.........7
.........8
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5         

metadata.json                                  2023-05-15 10:18:23           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0087 - accuracy: 0.9899 - val_loss: 0.0206 - val_accuracy: 0.9918
Epoch 69/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0084 - accuracy: 0.9888 - val_loss: 0.0265 - val_accuracy: 0.9796
Epoch 70/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0098 - accuracy: 0.9868 - val_loss: 0.0219 - val_accuracy: 0.9878
Epoch 71/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0094 - accuracy: 0.9878 - val_loss: 0.0226 - val_accuracy: 0.9837
Epoch 72/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0070 - accuracy: 0.9929 - val_loss: 0.0232 - val_accuracy: 0.9796
Epoch 73/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0068 - accuracy: 0.9909 - val_loss: 0.0217 - val_accuracy: 0.9796
Epoch 74/300
31/31 [==============================] - 0s 4ms/step - loss: 0.0102 

31/31 [==============================] - 0s 3ms/step - loss: 0.0170 - accuracy: 0.9777 - val_loss: 0.0720 - val_accuracy: 0.9592
Epoch 126/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0113 - accuracy: 0.9817 - val_loss: 0.0667 - val_accuracy: 0.9633
Epoch 127/300
31/31 [==============================] - 0s 2ms/step - loss: 0.0095 - accuracy: 0.9848 - val_loss: 0.0490 - val_accuracy: 0.9633
Epoch 128/300
31/31 [==============================] - 0s 2ms/step - loss: 0.0055 - accuracy: 0.9929 - val_loss: 0.0519 - val_accuracy: 0.9551
Epoch 129/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0034 - accuracy: 0.9959 - val_loss: 0.0473 - val_accuracy: 0.9714
Epoch 130/300
31/31 [==============================] - 0s 4ms/step - loss: 0.0033 - accuracy: 0.9959 - val_loss: 0.0481 - val_accuracy: 0.9633
Epoch 131/300
31/31 [==============================] - 0s 4ms/step - loss: 0.0032 - accuracy: 0.9990 - val_loss: 0.0496 - val_accuracy: 0.9673
Epoch 132/300

31/31 [==============================] - 0s 3ms/step - loss: 0.0033 - accuracy: 0.9959 - val_loss: 0.0636 - val_accuracy: 0.9592
Epoch 183/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0035 - accuracy: 0.9949 - val_loss: 0.0621 - val_accuracy: 0.9592
Epoch 184/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0045 - accuracy: 0.9959 - val_loss: 0.0971 - val_accuracy: 0.9592
Epoch 185/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0227 - accuracy: 0.9817 - val_loss: 0.0818 - val_accuracy: 0.9755
Epoch 186/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0091 - accuracy: 0.9868 - val_loss: 0.0617 - val_accuracy: 0.9714
Epoch 187/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0103 - accuracy: 0.9858 - val_loss: 0.0717 - val_accuracy: 0.9633
Epoch 188/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0121 - accuracy: 0.9858 - val_loss: 0.0682 - val_accuracy: 0.9673
Epoch 189/300

31/31 [==============================] - 0s 3ms/step - loss: 0.0365 - accuracy: 0.9665 - val_loss: 0.1139 - val_accuracy: 0.9551
Epoch 240/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0202 - accuracy: 0.9807 - val_loss: 0.1026 - val_accuracy: 0.9551
Epoch 241/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0245 - accuracy: 0.9767 - val_loss: 0.1019 - val_accuracy: 0.9551
Epoch 242/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0075 - accuracy: 0.9909 - val_loss: 0.1066 - val_accuracy: 0.9592
Epoch 243/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0043 - accuracy: 0.9970 - val_loss: 0.0832 - val_accuracy: 0.9592
Epoch 244/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0028 - accuracy: 0.9990 - val_loss: 0.0806 - val_accuracy: 0.9592
Epoch 245/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0025 - accuracy: 0.9990 - val_loss: 0.0779 - val_accuracy: 0.9592
Epoch 246/300

31/31 [==============================] - 0s 4ms/step - loss: 0.0014 - accuracy: 0.9990 - val_loss: 0.0907 - val_accuracy: 0.9633
Epoch 297/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0016 - accuracy: 0.9990 - val_loss: 0.0945 - val_accuracy: 0.9633
Epoch 298/300
31/31 [==============================] - 0s 4ms/step - loss: 0.0016 - accuracy: 0.9990 - val_loss: 0.0949 - val_accuracy: 0.9633
Epoch 299/300
31/31 [==============================] - 0s 4ms/step - loss: 0.0018 - accuracy: 0.9990 - val_loss: 0.0868 - val_accuracy: 0.9633
Epoch 300/300
31/31 [==============================] - 0s 4ms/step - loss: 0.0017 - accuracy: 0.9990 - val_loss: 0.0949 - val_accuracy: 0.9633


In [14]:
accurate_train_accuracy,classify_report= test_model(model,X_train,Y_train)

31/31 [==============================] - 0s 879us/step


In [15]:
accurate_train_accuracy

0.9989858012170385

In [16]:
print(classify_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00       127
           1       0.99      1.00      1.00       134
           2       1.00      1.00      1.00       271
           3       1.00      1.00      1.00       454

    accuracy                           1.00       986
   macro avg       1.00      1.00      1.00       986
weighted avg       1.00      1.00      1.00       986



In [17]:
accurate_val_accuracy,classify_report= test_model(model,X_val,Y_val)

8/8 [==============================] - 0s 1ms/step


In [18]:
accurate_val_accuracy

0.963265306122449

In [19]:
print(classify_report)

              precision    recall  f1-score   support

           0       0.96      0.86      0.91        29
           1       0.88      0.90      0.89        31
           2       0.96      0.99      0.97        73
           3       0.99      0.99      0.99       112

    accuracy                           0.96       245
   macro avg       0.95      0.94      0.94       245
weighted avg       0.96      0.96      0.96       245



In [20]:
accurate_test_accuracy,classify_report= test_model(model,X_test,Y_test)

8/8 [==============================] - 0s 1ms/step


In [21]:
accurate_test_accuracy

0.967479674796748

In [22]:
print(classify_report)

              precision    recall  f1-score   support

           0       0.89      0.94      0.91        34
           1       0.94      0.83      0.88        36
           2       0.97      1.00      0.98        62
           3       1.00      1.00      1.00       114

    accuracy                           0.97       246
   macro avg       0.95      0.94      0.95       246
weighted avg       0.97      0.97      0.97       246



## important-Note
gamma=1 on-test(0.967479674796748) on-val(0.9836734693877551)
gamma=2 on-test(0.967479674796748) on-val(0.9836734693877551) 
gamma=3 on-test(0.975609756097561) on-val(0.9836734693877551)

In [ ]:
## I can't consider observations in the  age to be false outlier 
## I will consider it a true outlier, I will just make a log transformation for it 
## bc it's so skewed , it have skewness value 1.481576 

# you can't do log transformation in the Time_E_DEV column 


# meal_Count log transforamtion has increase the skewness 





In [ ]:
df.skew().sort_values(ascending=False)

In [ ]:
df_temp= df.copy()

sns.boxplot(x=np.log(df_temp["Age"]))
plt.show()

df_temp[df_temp["Age"] < df_temp["Age"].quantile(0.25)]


plt.hist(df_temp["Age"])
plt.show()

In [ ]:
# pd.Series(np.log(df_temp["Time_E_Dev"])).skew()

df_temp=df.copy()

In [ ]:
plt.hist(np.log10(df_temp["Meal_Count"]))
plt.show()

In [ ]:
pd.Series(np.log10(df_temp["Meal_Count"])).skew()

In [ ]:
pd.Series(X_train[:,14]).skew()

In [ ]:
X_train[:,14]